### PR TITLE
Alternative solution using streaming-iterator-like approach

### PR DESCRIPTION
**Pull request by abigagli**
Hey, I'm trying to deepen my understanding of lifetimes as well, so as an exercise I created this alternative which basically mimics what a _streaming iterator_ probably would look like.
As far as I understand this cannot work with the current definition of the Iterator trait, at leas until GATs _(Generic Associated Types)_ will be suported (see e.g. [this post](https://lukaskalbertodt.github.io/2018/08/03/solving-the-generalized-streaming-iterator-problem-without-gats.html)).
So in this alternative solution we can't rely on just implementing the "official" iterator trait, and hence you can't just use for loops or iterator adapters.
I'd say it wouldn't be a viable solution for anything serious, but just wanted to test myself and contribute back to this nice repo.

EDIT: 
Also, I forgot to mention, each call to `next` would "lock" `self` until the result is dropped, since the returned value "borrows" from `self` and hence it requires for it to remain borrowed as long as the return value itself is alive.
Basically you could not call next multiple times and use the results of such calls together.
See [this stackoverflow answer](https://stackoverflow.com/a/40459231/41789) for a very similar scenario

### CODE REVIEW

**Comment by jayber** [comment]
Thank you so much for contributing to this repo!

It really warms my heart to see people make an effort to help and contribute, especially when the internet can seem so hostile.

If you have or encounter any good problems/solutions in your experiences, they can be added as whole new crates to this repo too, if you want.

Thanks again!
